modules/core/src/main/java/io/fluo/webindex/core/models/URL.java

```diff
@@ -0,0 +1,288 @@
+      return false;
+    }
+  }
+
+  public static boolean isValidIP(String host) {
+    return validator.isValid(host);
+  }
+
+  public static String reverseHost(String host) {
+    String[] hostArgs = host.split("\\.");
+    ArrayUtils.reverse(hostArgs);
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < hostArgs.length - 1; i++) {
+      sb.append(hostArgs[i]);
+      sb.append(".");
+    }
+    sb.append(hostArgs[hostArgs.length - 1]);
+    if (host.endsWith(".")) {
+      sb.append(".");
+    }
+    return sb.toString();
+  }
+
+  public boolean hasIPAddr() {
+    return isValidIP(host);
+  }
+
+  public String getHost() {
+    return host;
+  }
+
+  public String getReverseHost() {
+    if (hasIPAddr())
+      return host;
+    return reverseHost(host);
+  }
+
+  public String getPath() {
+    return path;
+  }
+
+  public boolean isSecure() {
+    return secure;
+  }
+
+  public int getPort() {
+    return port;
+  }
+
+  public boolean isImage() {
+    return path.matches("([^\\s]+(\\.(?i)(jpeg|jpg|png|gif|bmp))$)");
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder url = new StringBuilder();
+    url.append("http");
+    if (secure) {
+      url.append("s");
+    }
+    url.append("://");
+    url.append(host);
+    if (port != 80 && port != 443) {
+      url.append(":");
+      url.append(port);
+    }
+    url.append(path);
+    return url.toString();
+  }
+
+  public String toPageID() {
+    String reverseDomain = getReverseDomain();
+    String nonDomain = getReverseHost().substring(reverseDomain.length());
+    String portStr = "";
+    if ((!secure && port != 80) || (secure && port != 443)) {
+      portStr = Integer.toString(port);
+    }
+    return reverseDomain + ">" + nonDomain + ">" + (secure ? "s" : "o") + portStr + ">" + path;
+  }
+
+  public static URL fromPageID(String pageID) {
+    String[] idArgs = pageID.split(">");
+    if (idArgs.length != 4) {
+      throw new IllegalArgumentException("Page ID has too few or many parts: " + pageID);
+    }
+    String domain = reverseHost(idArgs[0]);
+    String host = reverseHost(idArgs[0] + idArgs[1]);
+    boolean secure = false;
+    int port = 80;
+    if (idArgs[2].startsWith("s")) {
+      secure = true;
+      port = 443;
+    } else if (!idArgs[2].startsWith("o")) {
+      throw new IllegalArgumentException("Page ID does not have port info beg with 's' or 'o': "
+          + pageID);
+    }
+    if (idArgs[2].length() > 1) {
+      port = Integer.parseInt(idArgs[2].substring(1));
+    }
+    String path = idArgs[3];
+    return new URL(domain, host, path, port, secure);
+  }
+
+  public String getDomain() {
+    return domain;
+  }
+
+  public String getReverseDomain() {
+    if (hasIPAddr()) {
+      return domain;
+    }
+    return reverseHost(domain);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
```

Comment:
could possibly use `instanceof` here... when null is passed to instanceof it returns false.